File: packages/dashcam_player/android/src/main/kotlin/com/dashcam/player/DashcamNativePlayer.kt
package com.dashcam.player

import android.os.Handler
import android.os.Looper
import android.view.SurfaceHolder
import android.view.SurfaceView
import kotlinx.coroutines.*
import java.net.HttpURLConnection
import java.net.URL

/**
 * FFmpeg-based RTSP player for F9 dashcam, adapted for Flutter plugin.
 *
 * Uses FFmpeg's native RTSP client with IPv4-only sockets,
 * bypassing Android's IPv6-first network stack.
 *
 * Events are sent to Flutter via the plugin's EventChannel.
 */
class DashcamNativePlayer(
    private val playerId: Int,
    private val plugin: DashcamPlayerPlugin
) {
    private val nativePlayer = NativeFFmpegPlayer()
    private var playerPtr: Long = 0
    private val scope = CoroutineScope(Dispatchers.IO + Job())
    private var heartbeatJob: Job? = null
    private var connectJob: Job? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private var connectionStartTime: Long = 0
    @Volatile private var isReleased = false
    @Volatile private var hasVideoRenderingStarted = false
    @Volatile private var currentCamera: Int = DashcamConfig.CAMERA_FRONT

    // SurfaceView reference — set when PlatformView is ready
    private var surfaceView: SurfaceView? = null

    // CompletableDeferred to await surface readiness before connect()
    private var surfaceReady = CompletableDeferred<Unit>()

    /**
     * Set the SurfaceView from the PlatformView.
     * Must be called before connect().
     */
    fun setSurfaceView(view: SurfaceView) {
        surfaceView = view
        setupSurface()
    }

    /**
     * Set up player surface with the SurfaceView's holder
     */
    private fun setupSurface() {
        val sv = surfaceView ?: return
        sv.holder.addCallback(object : SurfaceHolder.Callback {
            override fun surfaceCreated(holder: SurfaceHolder) {
                log("Surface created callback, initializing native player")
                initNativePlayer(holder)
            }

            override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
                log("Surface changed: ${width}x${height}")
            }

            override fun surfaceDestroyed(holder: SurfaceHolder) {
                log("Surface destroyed")
                // Reset for next surface creation cycle
                surfaceReady = CompletableDeferred()
            }
        })

        // Check if surface is ALREADY valid (callback was missed because
        // surfaceCreated fired before we registered the callback)
        if (sv.holder.surface != null && sv.holder.surface.isValid) {
            log("Surface already valid, initializing immediately (no callback)")
            initNativePlayer(sv.holder)
        }
    }

    /**
     * Initialize the native FFmpeg player with the given surface.
     * Called either from surfaceCreated callback or directly if surface was already ready.
     */
    private fun initNativePlayer(holder: SurfaceHolder) {
        if (playerPtr != 0L) {
            log("Native player already initialized, skipping")
            return
        }
        playerPtr = nativePlayer.nativeCreate(holder.surface)
        if (playerPtr == 0L) {
            log("ERROR: Failed to create native player")
            sendError("Failed to create native player")
        } else {
            log("Native player created: ptr=$playerPtr")
        }
        surfaceReady.complete(Unit)
    }

    /**
     * Connect to dashcam and start streaming.
     *
     * Sequence: ping → enterRecorder → getMediaInfo → heartbeat → startLive → waitPort → RTSP
     * Retries indefinitely until connected or cancelled.
     */
    suspend fun connect(cameraIndex: Int = DashcamConfig.CAMERA_FRONT): Boolean = withContext(Dispatchers.IO) {
        if (isReleased) {
            log("Cannot connect - player released")
            return@withContext false
        }

        currentCamera = cameraIndex

        // Wait for SurfaceView's surfaceCreated to fire and set playerPtr
        try {
            log("Waiting for surface to be ready...")
            withTimeout(5000L) {
                surfaceReady.await()
            }
            log("Surface is ready, playerPtr=$playerPtr")
        } catch (e: TimeoutCancellationException) {
            log("ERROR: Timed out waiting for surface (5s)")
            sendError("Surface not ready — timed out")
            return@withContext false
        }

        var attemptNumber = 0

        while (!isReleased && isActive) {
            attemptNumber++
            val verbose = attemptNumber <= 3 || attemptNumber % 5 == 0

            try {
                if (verbose) {
                    log("=== Connection attempt $attemptNumber ===")
                }
                sendStatus("Connecting... (attempt $attemptNumber)")

                // Step 0: Verify network connectivity
                if (verbose) log("Step 0: Verifying network connectivity...")
                val isReachable = pingDashcam()
                if (!isReachable && verbose) {
                    log("WARNING: Dashcam not reachable")
                }

                if (!isActive || isReleased) break

                // Step 1: Enter recorder mode
                if (verbose) log("Step 1: Enter recorder mode...")
                enterRecorderMode()

                // Step 2: Get media info
                if (verbose) log("Step 2: Get media info...")
                getMediaInfo()

                if (!isActive || isReleased) break

                // Step 3: Start heartbeat BEFORE RTSP
                if (verbose) log("Step 3: Starting heartbeat...")
                startHeartbeat()

                // Step 4: Start live preview (ACTIVATES the stream)
                if (verbose) log("Step 4: Start live preview - ACTIVATING STREAM...")
                sendStatus("Activating stream...")
                val liveStarted = startLivePreview(cameraIndex)
                if (!liveStarted && verbose) {
                    log("WARNING: Start live preview failed, continuing anyway...")
                }

                // Step 5: Wait for RTSP port
                if (verbose) log("Step 5: Waiting for RTSP port...")
                val isPortReady = waitForRtspPort(timeoutMs = 5000)
                if (!isPortReady && verbose) {
                    log("WARNING: RTSP port not responding, trying anyway...")
                }

                if (!isActive || isReleased) break

                // Step 6: Verify native FFmpeg loaded
                if (playerPtr == 0L) {
                    log("ERROR: Native player pointer is NULL")
                    stopHeartbeat()
                    sendError("Player not initialized - wait for surface")
                    return@withContext false
                }

                if (attemptNumber == 1) {
                    val testResult = nativePlayer.nativeTest()
                    log("Native FFmpeg test: $testResult")
                    if (!testResult.contains("working")) {
                        log("ERROR: FFmpeg in STUB mode!")
                        stopHeartbeat()
                        sendError("FFmpeg libraries not loaded")
                        return@withContext false
                    }
                }

                if (!isActive || isReleased) break

                // Step 7: Connect via FFmpeg RTSP (inner 3-retry)
                if (verbose) log("Step 7: Opening RTSP stream...")
                sendStatus("Opening RTSP stream...")
                connectionStartTime = System.currentTimeMillis()

                var connected = false
                for (rtspAttempt in 1..3) {
                    if (!isActive || isReleased) break
                    if (verbose) log("RTSP attempt $rtspAttempt/3")
                    connected = nativePlayer.nativeConnect(playerPtr, DashcamConfig.RTSP_URL)
                    if (connected) {
                        log("RTSP connected on attempt $rtspAttempt")
                        break
                    }
                    if (rtspAttempt < 3) {
                        delay(1000)
                    }
                }

                if (connected) {
                    // Start playback
                    log("Starting FFmpeg playback...")
                    nativePlayer.nativeStart(playerPtr)
                    log("FFmpeg playback started")

                    sendStatus("Playing")

                    if (!hasVideoRenderingStarted) {
                        hasVideoRenderingStarted = true
                        val latency = System.currentTimeMillis() - connectionStartTime
                        log("VIDEO STREAMING! Latency: ${latency}ms")
                        mainHandler.post {
                            plugin.sendEvent("videoRenderingStarted", mapOf("latencyMs" to latency.toInt()))
                        }
                    }

                    log("=== FFmpeg connection successful on attempt $attemptNumber ===")
                    return@withContext true
                }

                // Failed — stop heartbeat before retry
                stopHeartbeat()

            } catch (e: CancellationException) {
                log("Connection cancelled")
                stopHeartbeat()
                throw e
            } catch (e: Exception) {
                log("Attempt $attemptNumber failed: ${e.javaClass.simpleName} - ${e.message}")
                stopHeartbeat()
            }

            // Wait before next full attempt
            if (!isReleased && isActive) {
                log("Attempt $attemptNumber failed, retrying in 3s...")
                sendStatus("Retrying in 3s... (attempt $attemptNumber)")
                delay(3000)
            }
        }

        log("Connection loop stopped")
        false
    }

    /**
     * Switch camera via HTTP API and reconnect RTSP.
     *
     * Sequence: stop playback → HTTP switch API → wait for stream restart → reconnect RTSP
     */
    suspend fun switchCamera(camera: Int): Boolean = withContext(Dispatchers.IO) {
        try {
            log("Switching to camera: $camera")
            currentCamera = camera
            sendStatus("Switching camera...")

            // Stop current playback first
            if (playerPtr != 0L) {
                nativePlayer.nativeStop(playerPtr)
            }

            // Call HTTP API to switch camera on the dashcam
            val url = URL("${DashcamConfig.API_SWITCH_CAMERA}$camera")
            log("Camera switch API: $url")
            val switchOk = try {
                with(url.openConnection() as HttpURLConnection) {
                    requestMethod = "GET"
                    connectTimeout = 2000
                    readTimeout = 2000
                    setRequestProperty("User-Agent", DashcamConfig.USER_AGENT)
                    val code = responseCode
                    log("Camera switch API response: HTTP $code")
                    code == 200
                }
            } catch (e: Exception) {
                log("Camera switch API FAILED: ${e.javaClass.simpleName} - ${e.message}")
                false
            }

            if (!switchOk) {
                log("WARNING: Camera switch API failed, attempting reconnect anyway")
            }

            // Wait for RTSP stream to restart on new camera
            delay(500)
            val portReady = waitForRtspPort(timeoutMs = 5000)
            if (!portReady) {
                log("WARNING: RTSP port not ready after camera switch, trying anyway")
            }

            // Reconnect RTSP with infinite retry
            var rtspAttempt = 0
            var connected = false
            while (!isReleased && isActive) {
                rtspAttempt++
                log("Camera reconnect attempt $rtspAttempt")
                sendStatus("Reconnecting camera... (attempt $rtspAttempt)")
                connected = nativePlayer.nativeConnect(playerPtr, DashcamConfig.RTSP_URL)
                if (connected) {
                    log("Camera reconnect succeeded on attempt $rtspAttempt")
                    break
                }
                log("Camera reconnect failed, retrying in 3s...")
                delay(3000)
            }

            if (connected) {
                nativePlayer.nativeStart(playerPtr)
                sendStatus("Camera $camera active")
                log("Camera switch complete: camera $camera playing")
            } else {
                log("Camera switch cancelled")
                sendError("Camera switch cancelled")
            }
            connected

        } catch (e: Exception) {
            log("Camera switch error: ${e.javaClass.simpleName} - ${e.message}")
            sendError("Camera switch error: ${e.message}")
            false
        }
    }

    /**
     * Stop playback
     */
    fun stop() {
        try {
            if (!isReleased && playerPtr != 0L) {
                log("Stopping FFmpeg playback")
                connectJob?.cancel()
                stopHeartbeat()
                nativePlayer.nativeStop(playerPtr)
            }
        } catch (e: Exception) {
            log("Error stopping: ${e.message}")
        }
    }

    /**
     * Release all resources
     */
    fun release() {
        try {
            log("Releasing FFmpeg player")
            isReleased = true
            connectJob?.cancel()
            stopHeartbeat()
            // Cancel pending surface wait if any
            surfaceReady.cancel()
            if (playerPtr != 0L) {
                nativePlayer.nativeRelease(playerPtr)
                playerPtr = 0
            }
            scope.cancel()
        } catch (e: Exception) {
            log("Error releasing: ${e.message}")
        }
    }

    // === Private HTTP API Functions ===

    private suspend fun pingDashcam(): Boolean = withContext(Dispatchers.IO) {
        try {
            val url = URL(DashcamConfig.API_HEARTBEAT)
            log("Ping: $url")
            with(url.openConnection() as HttpURLConnection) {
                requestMethod = "GET"
                connectTimeout = 3000
                readTimeout = 3000
                val code = responseCode
                log("Ping response: HTTP $code")
                code == 200
            }
        } catch (e: Exception) {
            log("Ping FAILED: ${e.javaClass.simpleName} - ${e.message}")
            false
        }
    }

    private suspend fun enterRecorderMode(): Boolean = withContext(Dispatchers.IO) {
        try {
            val url = URL(DashcamConfig.API_ENTER_RECORDER)
            log("EnterRecorder: $url")
            with(url.openConnection() as HttpURLConnection) {
                requestMethod = "GET"
                connectTimeout = 2000
                readTimeout = 2000
                setRequestProperty("User-Agent", DashcamConfig.USER_AGENT)
                val code = responseCode
                log("EnterRecorder response: HTTP $code")
                code == 200
            }
        } catch (e: Exception) {
            log("EnterRecorder FAILED: ${e.javaClass.simpleName} - ${e.message}")
            false
        }
    }

    private suspend fun getMediaInfo(): Boolean = withContext(Dispatchers.IO) {
        try {
            val url = URL(DashcamConfig.API_GET_MEDIA_INFO)
            log("GetMediaInfo: $url")
            with(url.openConnection() as HttpURLConnection) {
                requestMethod = "GET"
                connectTimeout = 2000
                readTimeout = 2000
                setRequestProperty("User-Agent", DashcamConfig.USER_AGENT)
                val code = responseCode
                log("GetMediaInfo response: HTTP $code")
                code == 200
            }
        } catch (e: Exception) {
            log("GetMediaInfo FAILED: ${e.javaClass.simpleName} - ${e.message}")
            true // Non-critical
        }
    }

    private suspend fun startLivePreview(camIndex: Int = 0): Boolean = withContext(Dispatchers.IO) {
        try {
            val url = URL("${DashcamConfig.API_START_LIVE}$camIndex")
            log("StartLive: $url")
            with(url.openConnection() as HttpURLConnection) {
                requestMethod = "GET"
                connectTimeout = 2000
                readTimeout = 2000
                setRequestProperty("User-Agent", DashcamConfig.USER_AGENT)
                val code = responseCode
                log("StartLive response: HTTP $code")
                code == 200
            }
        } catch (e: Exception) {
            log("StartLive FAILED: ${e.javaClass.simpleName} - ${e.message}")
            false
        }
    }

    private suspend fun waitForRtspPort(timeoutMs: Long = 5000): Boolean = withContext(Dispatchers.IO) {
        val startTime = System.currentTimeMillis()
        var attempts = 0
        while (System.currentTimeMillis() - startTime < timeoutMs) {
            attempts++
            try {
                val socket = java.net.Socket()
                socket.connect(java.net.InetSocketAddress(DashcamConfig.DASHCAM_IP, DashcamConfig.RTSP_PORT), 500)
                socket.close()
                log("RTSP port ready after $attempts attempts")
                return@withContext true
            } catch (e: Exception) {
                delay(300)
            }
        }
        log("RTSP port not ready after $attempts attempts")
        false
    }

    private fun startHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = scope.launch {
            while (isActive && !isReleased) {
                delay(5000)
                try {
                    val url = URL(DashcamConfig.API_HEARTBEAT)
                    with(url.openConnection() as HttpURLConnection) {
                        requestMethod = "GET"
                        connectTimeout = 2000
                        readTimeout = 2000
                        if (responseCode == 200) {
                            log("Heartbeat: OK")
                        }
                    }
                } catch (e: Exception) {
                    log("Heartbeat: ${e.javaClass.simpleName}")
                }
            }
        }
    }

    private fun stopHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = null
    }

    // === Event Helpers ===

    private fun sendStatus(message: String) {
        mainHandler.post {
            plugin.sendEvent("statusChanged", mapOf("message" to message))
        }
    }

    private fun sendError(message: String) {
        mainHandler.post {
            plugin.sendEvent("error", mapOf("message" to message))
        }
    }

    private fun log(message: String) {
        android.util.Log.d("DashcamPlayer[$playerId]", message)
    }
}

File: packages/dashcam_player/ios/Classes/DashcamNativePlayer.swift
import Foundation

/// FFmpeg-based RTSP player for F9 dashcam on iOS.
///
/// Same 7-step connection sequence as Android:
/// ping → enterRecorder → getMediaInfo → heartbeat → startLive → waitPort → RTSP
class DashcamNativePlayer {

    private let playerId: Int
    private let plugin: DashcamPlayerPlugin
    private let bridge = DashcamNativeBridge()
    private var playerPtr: NSNumber?

    private var heartbeatTimer: Timer?
    private var isReleased = false
    private var shouldStopConnecting = false
    private var currentCamera: Int = DashcamConfig.CAMERA_FRONT

    init(playerId: Int, plugin: DashcamPlayerPlugin) {
        self.playerId = playerId
        self.plugin = plugin
    }

    // MARK: - Public API

    /// Set the MTKView for rendering. Called when PlatformView is ready.
    func setMetalView(_ view: UIView, platformView: DashcamPlatformView? = nil) {
        if playerPtr != nil {
            log("Native player already created, skipping")
            return
        }
        playerPtr = bridge.nativeCreate(with: view)
        if playerPtr?.int64Value == 0 {
            log("ERROR: Failed to create native player")
            sendError("Failed to create native player")
        } else {
            log("Native player created: ptr=\(playerPtr!)")

            // Pass bridge to PlatformView for MTKView delegate draw calls
            if let pv = platformView {
                pv.setBridge(bridge)
            }
        }
    }

    /// Connect to dashcam and start streaming.
    func connect(cameraIndex: Int = DashcamConfig.CAMERA_FRONT,
                 completion: @escaping (Bool) -> Void) {
        if isReleased {
            log("Cannot connect - player released")
            completion(false)
            return
        }

        // Signal any running connection/switch to stop
        shouldStopConnecting = true
        currentCamera = cameraIndex

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            // Wait for old connection/switch loop to exit
            Thread.sleep(forTimeInterval: 0.5)
            self.shouldStopConnecting = false
            let success = self.doConnect(cameraIndex: cameraIndex)
            DispatchQueue.main.async {
                completion(success)
            }
        }
    }

    /// Switch camera via HTTP API and reconnect RTSP.
    func switchCamera(camera: Int, completion: @escaping (Bool) -> Void) {
        // Signal any running connection/switch to stop
        shouldStopConnecting = true

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            // Wait for old connection loop to exit
            Thread.sleep(forTimeInterval: 0.5)
            self.shouldStopConnecting = false
            let success = self.doSwitchCamera(camera: camera)
            DispatchQueue.main.async {
                completion(success)
            }
        }
    }

    /// Stop playback.
    func stop() {
        guard !isReleased, let ptr = playerPtr, ptr.int64Value != 0 else { return }
        log("Stopping FFmpeg playback")
        shouldStopConnecting = true
        stopHeartbeat()
        bridge.nativeStop(ptr)
    }

    /// Release all resources.
    func release() {
        guard !isReleased else { return }
        log("Releasing FFmpeg player")
        isReleased = true
        shouldStopConnecting = true
        stopHeartbeat()
        if let ptr = playerPtr, ptr.int64Value != 0 {
            bridge.nativeRelease(ptr)
            playerPtr = nil
        }
    }

    // MARK: - Connection Sequence (same as Android)

    private func doConnect(cameraIndex: Int) -> Bool {
        log("=== Starting FFmpeg connection (camera=\(cameraIndex)) ===")

        var attemptNumber = 0

        while !isReleased && !shouldStopConnecting {
            attemptNumber += 1
            let verbose = attemptNumber <= 3 || attemptNumber % 5 == 0

            if verbose {
                log("=== Connection attempt \(attemptNumber) ===")
            }
            sendStatus("Connecting... (attempt \(attemptNumber))")

            // Step 0: Verify network
            if verbose { log("Step 0: Verifying network connectivity...") }
            let isReachable = pingDashcam()
            if !isReachable {
                if verbose { log("WARNING: Dashcam not reachable") }
            }

            if isReleased || shouldStopConnecting { break }

            // Step 1: Enter recorder mode
            if verbose { log("Step 1: Enter recorder mode...") }
            _ = enterRecorderMode()

            // Step 2: Get media info
            if verbose { log("Step 2: Get media info...") }
            _ = getMediaInfo()

            if isReleased || shouldStopConnecting { break }

            // Step 3: Start heartbeat
            if verbose { log("Step 3: Starting heartbeat...") }
            startHeartbeat()

            // Step 4: Start live preview
            if verbose { log("Step 4: Start live preview - ACTIVATING STREAM...") }
            sendStatus("Activating stream...")
            let liveStarted = startLivePreview(cameraIndex: cameraIndex)
            if !liveStarted && verbose {
                log("WARNING: Start live preview failed, continuing anyway...")
            }

            // Step 5: Wait for RTSP port
            if verbose { log("Step 5: Waiting for RTSP port...") }
            let portReady = waitForRtspPort(timeoutMs: 5000)
            if !portReady && verbose {
                log("WARNING: RTSP port not responding, trying anyway...")
            }

            if isReleased || shouldStopConnecting { break }

            // Step 6: Verify native player
            guard let ptr = playerPtr, ptr.int64Value != 0 else {
                log("ERROR: Native player pointer is NULL")
                stopHeartbeat()
                sendError("Player not initialized")
                return false
            }

            if attemptNumber == 1 {
                let testResult = bridge.nativeTest()
                log("Native FFmpeg test: \(testResult)")
                if !testResult.contains("working") {
                    log("ERROR: FFmpeg in STUB mode!")
                    stopHeartbeat()
                    sendError("FFmpeg libraries not loaded")
                    return false
                }
            }

            if isReleased || shouldStopConnecting { break }

            // Step 7: Connect via FFmpeg RTSP (inner 3-retry)
            if verbose { log("Step 7: Opening RTSP stream...") }
            sendStatus("Opening RTSP stream...")

            var connected = false
            for rtspAttempt in 1...3 {
                if isReleased || shouldStopConnecting { break }
                if verbose { log("RTSP attempt \(rtspAttempt)/3") }
                connected = bridge.nativeConnect(ptr, url: DashcamConfig.RTSP_URL)
                if connected {
                    log("RTSP connected on attempt \(rtspAttempt)")
                    break
                }
                if rtspAttempt < 3 {
                    Thread.sleep(forTimeInterval: 1.0)
                }
            }

            if connected {
                // Start playback
                log("Starting FFmpeg playback...")
                bridge.nativeStart(ptr)
                log("FFmpeg playback started")

                sendStatus("Playing")
                log("=== FFmpeg connection successful on attempt \(attemptNumber) ===")
                return true
            }

            // Failed — stop heartbeat before retry
            stopHeartbeat()

            // Wait 3 seconds before next full attempt (check cancellation every 0.5s)
            log("Attempt \(attemptNumber) failed, retrying in 3s...")
            sendStatus("Retrying in 3s... (attempt \(attemptNumber))")
            for _ in 0..<6 {
                if isReleased || shouldStopConnecting { return false }
                Thread.sleep(forTimeInterval: 0.5)
            }
        }

        log("Connection loop stopped")
        return false
    }

    private func doSwitchCamera(camera: Int) -> Bool {
        log("Switching to camera: \(camera)")
        currentCamera = camera
        sendStatus("Switching camera...")

        // Stop current playback
        if let ptr = playerPtr, ptr.int64Value != 0 {
            bridge.nativeStop(ptr)
        }

        // Call HTTP API to switch camera
        let urlString = "\(DashcamConfig.API_SWITCH_CAMERA)\(camera)"
        let switchOk = httpGet(urlString)
        if !switchOk {
            log("WARNING: Camera switch API failed, attempting reconnect anyway")
        }

        // Wait for stream restart
        Thread.sleep(forTimeInterval: 0.5)
        let portReady = waitForRtspPort(timeoutMs: 5000)
        if !portReady {
            log("WARNING: RTSP port not ready after camera switch")
        }

        // Reconnect RTSP with infinite retry
        guard let ptr = playerPtr, ptr.int64Value != 0 else {
            sendError("Player not available")
            return false
        }

        var attemptNumber = 0
        while !isReleased && !shouldStopConnecting {
            attemptNumber += 1
            let connected = bridge.nativeConnect(ptr, url: DashcamConfig.RTSP_URL)
            if connected {
                bridge.nativeStart(ptr)
                sendStatus("Camera \(camera) active")
                log("Camera switch complete: camera \(camera) playing")
                return true
            }

            log("Camera reconnect attempt \(attemptNumber) failed, retrying in 3s...")
            sendStatus("Reconnecting camera... (attempt \(attemptNumber))")
            for _ in 0..<6 {
                if isReleased || shouldStopConnecting { return false }
                Thread.sleep(forTimeInterval: 0.5)
            }
        }

        log("Camera switch cancelled")
        sendError("Camera switch cancelled")
        return false
    }

    // MARK: - HTTP API Helpers (same as Android)

    private func pingDashcam() -> Bool {
        return httpGet(DashcamConfig.API_HEARTBEAT)
    }

    private func enterRecorderMode() -> Bool {
        return httpGet(DashcamConfig.API_ENTER_RECORDER)
    }

    private func getMediaInfo() -> Bool {
        // Non-critical, return true even on failure
        _ = httpGet(DashcamConfig.API_GET_MEDIA_INFO)
        return true
    }

    private func startLivePreview(cameraIndex: Int) -> Bool {
        let url = "\(DashcamConfig.API_START_LIVE)\(cameraIndex)"
        return httpGet(url)
    }

    private func waitForRtspPort(timeoutMs: Int) -> Bool {
        let startTime = Date()
        var attempts = 0

        while Date().timeIntervalSince(startTime) < Double(timeoutMs) / 1000.0 {
            attempts += 1
            if checkTcpPort(host: DashcamConfig.DASHCAM_IP, port: DashcamConfig.RTSP_PORT, timeoutMs: 500) {
                log("RTSP port ready after \(attempts) attempts")
                return true
            }
            Thread.sleep(forTimeInterval: 0.3)
        }
        log("RTSP port not ready after \(attempts) attempts")
        return false
    }

    private func httpGet(_ urlString: String) -> Bool {
        guard let url = URL(string: urlString) else { return false }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 3.0
        request.setValue(DashcamConfig.USER_AGENT, forHTTPHeaderField: "User-Agent")

        let semaphore = DispatchSemaphore(value: 0)
        var success = false

        URLSession.shared.dataTask(with: request) { _, response, error in
            if let error = error {
                NSLog("DashcamPlayer[\(self.playerId)]: HTTP GET \(urlString) FAILED: \(error.localizedDescription)")
            } else if let httpResponse = response as? HTTPURLResponse {
                success = httpResponse.statusCode == 200
                NSLog("DashcamPlayer[\(self.playerId)]: HTTP GET \(urlString) → \(httpResponse.statusCode)")
            }
            semaphore.signal()
        }.resume()

        semaphore.wait()
        return success
    }

    // MARK: - Heartbeat

    private func startHeartbeat() {
        stopHeartbeat()
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { [weak self] _ in
            guard let self = self, !self.isReleased else { return }
            self.httpGet(DashcamConfig.API_HEARTBEAT)
        }
    }

    private func stopHeartbeat() {
        heartbeatTimer?.invalidate()
        heartbeatTimer = nil
    }

    // MARK: - Event Helpers

    private func sendStatus(_ message: String) {
        DispatchQueue.main.async {
            self.plugin.sendEvent("statusChanged", data: ["message": message])
        }
    }

    private func sendError(_ message: String) {
        DispatchQueue.main.async {
            self.plugin.sendEvent("error", data: ["message": message])
        }
    }

    private func log(_ message: String) {
        NSLog("DashcamPlayer[\(playerId)]: \(message)")
    }
}

// MARK: - Simple TCP Socket Helper

private func checkTcpPort(host: String, port: Int, timeoutMs: Int) -> Bool {
    var addr = sockaddr_in()
    addr.sin_family = sa_family_t(AF_INET)
    addr.sin_port = in_port_t(port).bigEndian
    inet_pton(AF_INET, host, &addr.sin_addr)

    let sock = Darwin.socket(AF_INET, SOCK_STREAM, 0)
    if sock < 0 { return false }
    defer { Darwin.close(sock) }

    // Set send timeout for connect
    var tv = timeval(tv_sec: timeoutMs / 1000, tv_usec: Int32((timeoutMs % 1000) * 1000))
    setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))

    let result = withUnsafePointer(to: &addr) { ptr in
        Darwin.connect(sock, UnsafeRawPointer(ptr).assumingMemoryBound(to: sockaddr.self), socklen_t(MemoryLayout<sockaddr_in>.size))
    }

    return result == 0
}
